1st assigment

Kubci opened the pull request:
+Tire entity specified
+Dao Interface specified
+Dao Interface crud implemented 
+In Memory derby database added

Kubci committed:
1st assigment

Kubci committed:
bad copy deleted

Kubci committed:
adding references to db

Kubci committed:
rename

Kubci committed:
rename

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/entity/tire/TireManagerDao.java:
```diff
@@ -0,0 +1,73 @@
+ * TireProperties are kept in db. It is expected that these values will be reused 
+ * and are changing rarely.
+ */
+public interface TireManagerDao {
```
This should be in a package .backend.dao, not in .backend.entity.

tinnyx commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/entity/tire/SpeedClass.java:
```diff
@@ -0,0 +1,37 @@
+    W(270),
+    Y(300);
+    
+    private final int id;
```
This should be called 'speed' instead of 'id', to be more futureproof.

martinzilak replied:
This should be called 'speed' instead of 'id', to be more futureproof.

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/entity/tire/TireManagerDaoImpl.java:
```diff
@@ -0,0 +1,156 @@
+ *
+ * @author Jakub Palenik, 422453@mail.muni.cz
+ */
+public class TireManagerDaoImpl implements TireManagerDao {
```
This should be in a package .backend.dao, not in .backend.entity.

Kubci committed:
refractor - package for dao's added

Kubci committed:
refractor - name change

Kubci committed:
refractor - enums added to enum names, moved to separate package

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/dao/TireManagerDao.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.List;
+
+/***
+ * @author Jakub Palenik, 422453@mail.muni.cz
```
Author should be under the class description itself.

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/dao/TireManagerDao.java:
```diff
@@ -0,0 +1,79 @@
+     * @return true on success
+     * @throws IllegalArgumentException on any param == null
+     */
+    boolean updateTire(Tire old, Tire data) throws IllegalArgumentException;
```
Update should take just one parameter which is of the type Tire that already has the id set. Update method then looks for a Tire with this id and updates it accordingly.

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/dao/TireManagerDao.java:
```diff
@@ -0,0 +1,79 @@
+     * @param filter
+     * @return true on success
+     */
+    List<Tire> getTiresByFilter(Tire filter);
```
This probably requires a new class that manages the filtering logic. As it is, it wouldn't allow to search for Tires by a field that is null. Let's imagine that I need to want for all Tires that don't have the Manufacturer set, which this wouldn't allow me to.

I suggest creating a class TireFilter that has all the fields that Tire does, but has them as Optional. Then, Optional.empty() can be used to look for fields that are supposed to be null. This example should be much more explanatory:

```java
public class TireManagerDaoImpl {
  public List<Tire> getTiresByFilter(TireFilter filter) {
    SqlStatement sql = sql(); // Just an example of SQL
    // Filtering a field that can't be null
    if (filter.id != null) {
      sql.filter("id", filter.id);
    }
    // Filtering a field that can be null
    if (filter.manufacturer != null) {
      sql.filter("manufacturer", filter.manufacturer.orElse(null));
    }
  }
  
  public class TireFilter {
    private Long id;
    // Manufacturer does not have to be set for a tire
    private Optional<Manufacturer> manufacturer;
  }
  
  public void main(String... args) {
    Filter f = new TireFilter();
    f.manufacturer = Optional.empty();
    // Finds all tires where the manufacturer is not set
    new TireManagerDaoImpl().getTiresByFilter(filter);
  }
}
```

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/entity/tire/Manufacturer.java:
```diff
@@ -0,0 +1,107 @@
+ * @author Jakub Palenik, 422453@mail.muni.cz
+ */
+@Entity
+public class Manufacturer {
```
Should be called TireManufacturer.

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/entity/tire/Manufacturer.java:
```diff
@@ -0,0 +1,107 @@
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
```
Should use instanceof instead because of subclassing.

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/entity/tire/TireProperties.java:
```diff
@@ -0,0 +1,238 @@
+ * @author Jakub Palenik, 422453@mail.muni.cz
+ */
+@Entity
+public class TireProperties {
```
This shouldn't be separate. These fields should be inside of Tire class.

Kubci replied:
This should be separate, otherwise there would be redundant data in table Tire.  One tire property  has multiple different tires.

martinzilak replied:
It's higly improbable that there'd be multiple tires from different manufacturers with the exact same properties. I wouldn't consider it redundancy if some subset of the fields repeats in multiple tires.

Kubci replied:
On the contrary it is higly probable. Because to maximize your profit as a company you want to sell pneus for most used tire sizes.

martinzilak commented on backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/entity/tire/Tire.java:
```diff
@@ -0,0 +1,164 @@
+    @DecimalMin("0.0")
+    private BigDecimal price;
+    
+    private String image_url;
```
Should be imageUrl instead.

Kubci committed:
+comments adjusted
+updateTire method to take only one param
+filter removed
+Manufacturer renamed
+getClass onto instanceof changed
+imge_url onto imageUrl renamed

Kubci committed:
tryout out
Please enter the commit message for your changes. Lines starting

Kubci committed:
licence bs removed

Kubci committed:
licence bs removed
